### ISSUE TITLE
accuracy calculation

### ISSUE THREAD

**Issue by PC09** [UKPLab/emnlp2017-relation-extraction#14]
I have trained the ContextWeighted model on given Wikidata json data. While training, the batch accuracy is ~30% while after training and saving the weights, the evaluate function calculates the accuracy which comes out to be ~94% . Can you please explain why this difference is coming and how accuracy is calculated in evaluate function?
PFA the ouput logs-
Ouptut when model is getting trained(last epoch)-
372608/372877 [============================>.] - ETA: 2s - loss: 0.0671 - acc: 0.2792
372736/372877 [============================>.] - ETA: 1s - loss: 0.0671 - acc: 0.2792
372864/372877 [============================>.] - ETA: 0s - loss: 0.0671 - acc: 0.2792
372877/372877 [==============================] - 3357s 9ms/step - loss: 0.0671 - acc: 0.2792 - val_loss: 0.1212 - val_acc: 0.2613

Epoch 00015: val_loss did not improve
Epoch 00015: early stopping

Ouptut from evaluate function-
Results (per sentence):  0.006015388452492377
Results: Accuracy:  0.9474328692033872
Results: Micro-Average F1:  (0.9424859243294846, 0.9418660537794111, 0.9421758870991064)

Also, why val_loss is used as a monitoring parameter? Can we take training_loss , will that be better ?

Thanks,
Parul

**Comment by daniilsorokin** [comment]
Hi,

super sorry for the late reply. 

The evaluate function accuracy is the correct one and 94% on the training set is what you should expect. The reason it is reported as being only 0.279 during training is because of the masking of the output. The model tries to predict all relations in the same sentence jointly, but since the number of relations is always different the output labels are padded with zeros to the maximum amount of 7. The loss function is implemented here so that those zeros are ignored (since they are meaningless), but the keras accuracy function doesn't know that and computes accuracy over the 7 predicted relations. One could re-implement the accuracy function as well, but it is not necessary for a successful training. 

As to your second question, the val_loss is monitored for the early stopping. It makes for a model that can better generalise. The val_loss is computed over the validation data set that is different from the held out (test) data.

I hope it helps. Feel free to follow up if not!

Best,
Daniil
